In [ ]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, SGDRegressor
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

In [ ]:
df = pd.read_csv('/content/advertising.csv')
df.head()

,TV,Radio,Newspaper,Sales
0,230.1,37.8,69.2,22.1
1,44.5,39.3,45.1,10.4
2,17.2,45.9,69.3,12.0
3,151.5,41.3,58.5,16.5
4,180.8,10.8,58.4,17.9


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   TV         200 non-null    float64
 1   Radio      200 non-null    float64
 2   Newspaper  200 non-null    float64
 3   Sales      200 non-null    float64
dtypes: float64(4)
memory usage: 6.4 KB


In [ ]:
X = df.drop('Sales', axis=1)
y = df['Sales']

linear regression


In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

In [ ]:
lr = LinearRegression()

In [ ]:
lr.fit(X_train, y_train)

LinearRegression()

In [ ]:
lr_pred = lr.predict(X_test)

In [ ]:
print("R2 Score:", r2_score(y_test, lr_pred))
print("MAE:", mean_absolute_error(y_test, lr_pred))
print("RMSE:", np.sqrt(mean_squared_error(y_test, lr_pred)))

R2 Score: 0.9059011844150826
MAE: 1.2748262109549338
RMSE: 1.7052146229349223


In [ ]:
print("Coefficient:", lr.coef_)
print("Intercept:", lr.intercept_)

Coefficient: [0.05450927 0.10094536 0.00433665]
Intercept: 4.714126402214127


SGD Regressor

In [ ]:
sgd = SGDRegressor(
    max_iter=1000,
    eta0=0.01,
    random_state=42
)

In [ ]:
sgd.fit(X_train, y_train)

SGDRegressor(random_state=42)

In [ ]:
sgd_pred = sgd.predict(X_test)

In [ ]:
print("R2 Score:", r2_score(y_test, sgd_pred))
print("MAE:", mean_absolute_error(y_test, sgd_pred))
print("RMSE:", np.sqrt(mean_squared_error(y_test, sgd_pred)))

R2 Score: -1.805188697149279e+24
MAE: 6742569747555.786
RMSE: 7468756017535.958


In [ ]:
print("Coefficient:", sgd.coef_)
print("Intercept:", sgd.intercept_)

Coefficient: [2.99420951e+10 9.66937850e+10 6.49567285e+09]
Intercept: [8.23707133e+10]


In [ ]:
print("Linear Regression R2 Score :", r2_score(y_test, lr_pred))
print("SGD Regressor R2 Score      :", r2_score(y_test, sgd_pred))

Linear Regression R2 Score : 0.9059011844150826
SGD Regressor R2 Score      : -1.805188697149279e+24


Linear Regression performed very well on the dataset.

It explained about 90.59% of the variation in the sales data.

SGD Regressor performed very poorly with the current settings.

The negative R² score means the SGD model made very inaccurate predictions.

This happened because SGD requires feature scaling (StandardScaler) before training.

Therefore, Linear Regression is the better model for this dataset.